### PR TITLE
avfilter: add QSV variants of the stack filters

### PR DESCRIPTION
**Pull request by xhaihao**
Some code is copy and pasted from other filters

Example:
ffmpeg -hwaccel qsv -c:v hevc_qsv -i input.h265 -filter_complex
"[0:v][0:v]hstack_qsv" -f null -

### CODE REVIEW

File: libavfilter/vf_stack_qsv.c [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+    int nb_inputs;
+    int shortest;
+    double scale;
+    char *layout;
+    uint8_t fillcolor[4];
+    char *fillcolor_str;
+    int fillcolor_enable;
+} QSVStackContext;
+
+static void rgb2yuv(float r, float g, float b, int *y, int *u, int *v, int depth)
```
Comment (uartie) [inline libavfilter/vf_stack_qsv.c:71]: Is there a common definition in ffmpeg for rgb2yuv?

**Review by uartie — COMMENTED** [review]
What do stack filters do?

**Comment by xhaihao** [comment]
> What do stack filters do?

See https://ffmpeg.org/ffmpeg-filters.html#toc-hstack, 

"Note that this filter is faster than using overlay and pad filter to create same output. "

 https://ffmpeg.org/ffmpeg-filters.html#toc-vstack  and https://ffmpeg.org/ffmpeg-filters.html#xstack

File: libavfilter/vf_stack_qsv.c (hunk 1 above)
Comment (xhaihao) [reply libavfilter/vf_stack_qsv.c:71]: I see a few different implementations of rgb2yuv in FFmpeg, the above one was copy and pasted from vf_colorize.c

**Comment by xhaihao** [comment]
there is a fate error:
- /opt/media/src/assets/otc-media/fate/apng/alogo.png: No such file or directory

It seems we should update fate clips.

**Comment by uartie** [comment]
> there is a fate error:
> 
> * /opt/media/src/assets/otc-media/fate/apng/alogo.png: No such file or directory
> 
> It seems we should update fate clips.

It looks like the chosen CI test agent did not have the asset sync service running.  I have restarted it and retrigger the checks.

**Comment by xhaihao** [comment]
Any other comment or objection? If no, I will send this patch to FFmpeg mailing list.

File: libavfilter/vf_stack_qsv.c [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+    QSVStackContext *sctx = ctx->priv;
+    AVFilterLink *inlink0 = ctx->inputs[0];
+    int width, height, i, ret;
+    enum AVPixelFormat in_format;
+    int Y, U, V, depth = 8;
+
+    if (inlink0->format == AV_PIX_FMT_QSV) {
+         if (!inlink0->hw_frames_ctx || !inlink0->hw_frames_ctx->data)
+             return AVERROR(EINVAL);
+         else
```
Comment (xuguangxin) [inline libavfilter/vf_stack_qsv.c]: else is not needed

File: libavfilter/vf_stack_qsv.c [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+
+    for (i = 1; i < sctx->nb_inputs; i++) {
+        AVFilterLink *inlink = ctx->inputs[i];
+
+        if (inlink0->format == AV_PIX_FMT_QSV) {
+            if (inlink0->format != inlink->format) {
+                av_log(ctx, AV_LOG_ERROR, "Mixing hardware and software pixel formats is not supported.\n");
+
+                return AVERROR(EINVAL);
+            } else {
```
Comment (xuguangxin) [inline libavfilter/vf_stack_qsv.c]: else is not needed.

File: libavfilter/vf_stack_qsv.c [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+        for (i = 0; i < sctx->nb_inputs; i++) {
+            AVFilterLink *inlink = ctx->inputs[i];
+            mfxVPPCompInputStream *is = &sctx->comp_conf.InputStream[i];
+
+            if (inlink0->w != inlink->w) {
+                av_log(ctx, AV_LOG_ERROR, "Input %d width %d does not match input %d width %d.\n", i, inlink->w, 0, inlink0->w);
+                return AVERROR(EINVAL);
+            }
+
+            is->DstX = 0;
```
Comment (xuguangxin) [inline libavfilter/vf_stack_qsv.c]: maybe you can use a function for this.
set_input_param(is, x, y, w, h)

File: libavfilter/vf_stack_qsv.c (hunk 2 above)
Comment (xhaihao) [reply libavfilter/vf_stack_qsv.c]: Updated, thank!

File: libavfilter/vf_stack_qsv.c (hunk 3 above)
Comment (xhaihao) [reply libavfilter/vf_stack_qsv.c]: Updated, thank!

File: libavfilter/vf_stack_qsv.c (hunk 4 above)
Comment (xhaihao) [reply libavfilter/vf_stack_qsv.c]: Defined a MACRO to set the parameters,  thanks

**Review by xuguangxin — APPROVED** [review]

**Comment by xhaihao** [comment]
patch was sent to FFmpeg mailing list.
http://ffmpeg.org/pipermail/ffmpeg-devel/2021-June/281252.html

**Pull request closed without merging**